Cython/Build/Tests/TestCyCache.py

```diff
@@ -23,38 +23,34 @@ def cache_files(self, file_glob):
         return glob.glob(os.path.join(self.cache_dir, file_glob))
 
     def fresh_cythonize(self, *args, **kwargs):
-        Cython.Utils.clear_function_caches()
-        Cython.Build.Dependencies._dep_tree = None  # discard method caches
-        Cython.Build.Dependencies.cythonize(*args, **kwargs)
+        kwargs.update(cache=self.cache_dir)
+        fresh_cythonize(*args, **kwargs)
 
     def test_cycache_switch(self):
         content1 = 'value = 1\n'
         content2 = 'value = 2\n'
         a_pyx = os.path.join(self.src_dir, 'a.pyx')
         a_c = a_pyx[:-4] + '.c'
 
-        with open(a_pyx, 'w') as f:
-            f.write(content1)
-        self.fresh_cythonize(a_pyx, cache=self.cache_dir)
-        self.fresh_cythonize(a_pyx, cache=self.cache_dir)
+        write_file(a_pyx, content1)
+        self.fresh_cythonize(a_pyx)
+        self.fresh_cythonize(a_pyx)
         self.assertEqual(1, len(self.cache_files('a.c*')))
         with open(a_c) as f:
             a_contents1 = f.read()
         os.unlink(a_c)
 
-        with open(a_pyx, 'w') as f:
-            f.write(content2)
-        self.fresh_cythonize(a_pyx, cache=self.cache_dir)
+        write_file(a_pyx, content2)
+        self.fresh_cythonize(a_pyx)
         with open(a_c) as f:
             a_contents2 = f.read()
         os.unlink(a_c)
 
         self.assertNotEqual(a_contents1, a_contents2, 'C file not changed!')
```

Comment:
For this case, I think to move `relative_lines_from_file` to `TestUtils`, only then either change the function, because it is suitable for specific cases, or change the name.

Reply:
Seems a reasonable addition there, but I repeat my suggestion from https://github.com/cython/cython/pull/4241#discussion_r656600320

Reply:
I'm talking about a little different, about the `except` part, there is a search for `VARS_LINE` and `lines[ind-10: ind-1]`, which fits well that test case, but not all, it was assumed for those tests that all important parts in the case the failure will be in `VARS_LINE` area.

Reply:
There is an idea to do this: take out part of `try` into a helper function, and then use it in test cases with a special modification for them.
Or add something like "if not found" kw argument, though we need to pass a string around which we should search if not found, and also what lines around we should present in an error.
Maybe it can be a function that will return all this, or rather a tuple .. oh, how many ideas :)

Reply:
Sketched out [here](https://github.com/cython/cython/pull/4241/commits/d9ab66765f5e6207b60ce9ff1716d4e88d75c0f0), it remains to add tests and use

Reply:
used [here](https://github.com/cython/cython/pull/4241/commits/6cebfd32cc9f3b462aad5528a03057b3eb194213) and [here](https://github.com/cython/cython/pull/4241/commits/0ea7ed167d1935d87b39cda0ba1f3ba6f07fcebe), remains to add tests

Reply:
[enough](https://github.com/cython/cython/pull/4241#issuecomment-878709162) tests for this pr
basically test are [here](https://github.com/cython/cython/pull/4241/commits/35035f363004fee16314e88c20dc68deafc636f2), [here](https://github.com/cython/cython/pull/4241/commits/14079b723b2baaea79cd417ce53eee9c51da072b), [here](https://github.com/cython/cython/pull/4241/commits/b3eecfbbeb632977525453fe5dbb323a05e281e6), [here](https://github.com/cython/cython/pull/4241/commits/6ea3c569dd1482521fff9b58e76ac2e314a588cd), and [here](https://github.com/cython/cython/pull/4241/commits/1edc3543105beaa2e7d3e15b63131ab7d269d81a)

Reply:
and [here](https://github.com/cython/cython/pull/4241/commits/03c3668a0e7524146d058ae410df29542b166334) are the tests

finally resolved